### PR TITLE
[crypto] Simplify `VerificationObligation`.

### PR DESCRIPTION
**Pull request by huitseeker**
The main contribution it the last commit (the first is straightforward).

## Context
The `VerificationObligation` is an accumulator for **batch verification of signatures on the same message**. It works predicated on the `batch_verify` method of the `VerifyingKey` trait, which as we know works in this way:
- performs a batch verification on these schemes that can (Ed25519, BLS),
- performs an iterated verification on these schemes that can't (Secp256k1)

## Issue
The trait `ToObligationSignature`, despite comments to the contrary, was doing runtime detection of the argument's type in the call to the `VerificationObligation::add_signature_and_public_key`, which took any signature as argument, converted some of them to an obligation (for BLS), the others to an `Obligation::None` (for the other two schemes, despite Ed25519 having a valid batch verification semantic), and errored upon encountering any `Obligation::None`.

The reason for this behavior isn't that we want to address verification generically somehow, it is that we want to build accumulators (`VerificationObligation`) for the authority signatures: if you happen to want to add a signature to the batch that's not of the authority's scheme (in our case, BLS), that's an error.

## This PR
A simple way of detecting this statically is to make `VerificationObligation` monomorphic, making calling `add_signature_and_public_key` on anything but the authority's chosen scheme a compilation error. That's what this commit does.

### CODE REVIEW

File: crates/sui-types/src/crypto.rs [hunk 1]
```diff
@@ -965,25 +958,6 @@ impl<S: SuiSignatureInner + Sized> SuiSignature for S {
                 error: format!("{}", e),
             })
     }
-
-    fn add_to_verification_obligation_or_verify(
```
Comment (kchalkias) [inline crates/sui-types/src/crypto.rs:969]: So this function will either add to the obligation list (anything that can be batched verified) or literally verify the signature. Now that we removed it, where do we verify sigs that cannot be batched?

Comment (huitseeker) [reply crates/sui-types/src/crypto.rs:969]: This function was 100% unused, so there was no change in behaviour related to this function.
The place where we check the user's sig is here:
https://github.com/MystenLabs/sui/blob/97e96ba1cf2cd8226bc31a7aecd35b6fdd7ea37a/crates/sui-core/src/consensus_validator.rs#L72

**Review by kchalkias — APPROVED** [review]
Thanks for pointing out the function was unused anyway.

File: crates/sui-types/src/crypto.rs [hunk 2]
```diff
@@ -1458,32 +1399,24 @@ impl VerificationObligation {
...
-            ObligationSignature::None => Err(SuiError::SenderSigUnbatchable),
-        }
+        self.public_keys
+            .get_mut(idx)
+            .ok_or(SuiError::InvalidAuthenticator)?
+            .push(public_key.clone());
+        self.signatures
+            .get_mut(idx)
+            .ok_or(SuiError::InvalidAuthenticator)?
+            .add_signature(signature.clone())
```
Comment (joyqvq) [inline crates/sui-types/src/crypto.rs:1415]: is this way to get rid of the clone here?

Comment (huitseeker) [reply crates/sui-types/src/crypto.rs:1415]: By changing the function's signature to take ownership of its argument.

**Pull request merged**